### ISSUE TITLE
[FR] Future idea for autodeploys and versioning

### ISSUE THREAD

**Issue by johnny-bit** [darktable-org/dtdocs#76]
I had a thought recently when looking at old darktable manuals and other manuals - the problem of versions. This is just to start discussion but here's my thought:

Currently dtdocs is just "master" but I believe in future it'll have more versions. My idea is that master should always follow darktable master and there should be specific version-branches to have some kind of rewording/spelling etc open instead of fixed tags.
On autodeployed site the pages should be versioned and  the "default" route should direct to latest stable version.
Last part: help items inside darktable and their generation: darktable knows which version it is (like for example db backups) so calling help items in form of dtdocs/[VERSION]/page. Any "unknown" version could be interpreted as "master" or anything like that.

Obvously the last part requires buy-in from darktable side, but it's imo worth pursuing

**Comment by elstoc** [comment]
Yes I was thinking along similar lines. Sounds reasonable.

**Comment by elstoc** [comment]
One thing we'll need to consider is that we can only have one gh-pages site per repository so publishing multiple versions might need some thought. I'd like to branch this for 3.4 before we start incorporating features from the next dev cycle so I'll hold off merging PRs documenting those features until we've agreed on an approach.

**Comment by paperdigits** [comment]
So I've been thinking about this, and I really want to get the documentation published with the rest of darktable.org. 

I have a feature request open to port the main website to hugo. I am going to start working on this soon. Once that is accepted, hugo has a "mount" feature that is basically the same idea as a bind mount. We're using it to get font awesome in the right place in our config.yaml:
```
module:
  mounts:
    - source: ./themes//hugo-darktable-docs-theme/assets/node_modules/fork-awesome/fonts
      target: static/fonts
```
So, in the main site, for each docs release (3.4, 3.6, etc etc), we can check out a specific branch/tag into a folder at the root of the dtorg site, then we can mount the content folder to the correct location under dtorg's content folder. So the URL for the documentation darktable.org/usermanual/3.4/index.html

**Comment by elstoc** [comment]
v3.6 is the point at which we will need to deal with maintaining the link between darktable versions and dtdocs versions (since the help links will be updated by then). So I suggest we wait until we have a translation workflow in place and the migration of darktable.org has settled down and then agree how to manage this.

**Comment by elstoc** [comment]
I would prefer the following scheme...

- The development version automatically deploys to `darktable.org/usermanual/dev/` and odd-numbered versions of darktable have help links that point to this location
- The latest stable version is deployed to `darktable.org/usermanual/latest/`
- The help links for the latest stable version point to `darktable.org/usermanual/3.6/` (for example) and, if 3.6 is the latest stable version, this automatically redirects to `darktable.org/usermanual/latest/`. This means that if you're running the current version, the links always point to a semi-permanent location that you can share with others.
- Once a new version is released, the previous version is moved from `darktable.org/usermanual/latest/` to `darktable.org/usermanual/3.6/` (for example) to replace the previous redirect. If possible these older versions could also contain a warning that a new version is available (and a link).

@TurboGit, @paperdigits what do you think? Ideally we should try to get this up and running before the v3.6 freeze so that we can test the help links.

**Comment by johnny-bit** [comment]
I'm not sure whether it should be `/3.6` -> `/latest` or `/latest` -> `/3.6` in terms of redirects but overall sounds good.

**Comment by elstoc** [comment]
I'm trying to keep it so that, as much as possible, people only see the `/latest` site and we don't have lots of links to old versions hanging around on the internet.

**Comment by TurboGit** [comment]
For the integrated help we will need to deal with:
- `/dev`
- `/<ver>`

So the `/latest` is not necessary for this part. But overall I agree with the proposal. Let's go with that to be able to implement this in dt and test before the release.

**Comment by elstoc** [comment]
@TurboGit agreed - you shouldn't need to know about `/latest` at all for the help integration, whichever solution we go for with darktable.org.

**Comment by johnny-bit** [comment]
OK, I had further thoughts about that...
So generally I now agree that newest stable version should redirect to latest (302 redirect). some seo/schema.org/canonicalization etc would be needed at some point but this can be done waaay later :)

**Comment by paperdigits** [comment]
Are we still planning to branch on a release version? If so, we'll nee to branch, then get that branch into weblate to continue translation, though it doesn't look like any of the translations will be close to complete for 3.6.

**Comment by elstoc** [comment]
For this release, we'll just tag it and release as English-only. We'll have to discuss, for 3.8, whether we're going to ask for translations to be finished when we release or whether we're going to continue to update in a 3.8 branch. Our current release plan is that we won't release new features except in our twice-yearly releases, so there's no need to branch for the English version.

**Comment by elstoc** [comment]
We're going with `/stable` as the 'permanent' URL. So for version 3.6 darktable helplinks will point to `darktable.org/usermanual/3.6`, which will redirect to `darktable.org/usermanual/stable`. When we release version 3.8 then the helplinks will point to `darktable.org/usermanual/3.8` and redirect to `/stable`. People still using version 3.6 at that point will link directly to an archived copy at `/3.6`. Hope that's clear!

**Comment by elstoc** [comment]
Ok we now have stuff published to darktable.org but we're having some discussions around all the required locations/redirects so I thought I'd do a new summary detailing the outcome of those discussions.

All of the dtdocs sites have a url constructed as follows `docs.darktable.org/usermanual/<ver>[/<lang>][/path/to/page/]`. Note that the parts in the square brackets are essentially optional and some additional nginx config will need to handle this.

### versions

The HEAD of dtdocs current master should always be published to `<ver>=development`. Help links from darktable master will always point to this location.

Stable versions are a little more complicated, since darktable will always include the version _number_ in the URL of help links. If that number is the latest stable version we want `<ver>=3.6`, for example, to redirect to `<ver>=stable` on darktable.org. Older stable versions should include the version number in their URL (e.g. `<ver>=3.4`) and not redirect further. In dtdocs stable versions will be denoted by tags X.Y.Z. Stable will always be the largest value of X.Y, and Z denotes the most recent dtdocs release of that version. These tags may be against commits in the master branch or against release-specific branches (if master has moved on and we need to update an older version).

The main documents landing page (docs.darktable.org) should have links to development, stable, and all previous available versions. It would be great if we could republish the 3.4 version of dtdocs.

### languages

There is currently a mis-match between how darktable and dtdocs denote available language codes. This will be changed in dtdocs so that language codes are consistent. We will need to back-port this to 3.6.

For development, all languages will be included (including partial translations). It is still to be determined whether we do this for stable versions - we used to only include complete translations.

Each language will be in its own subdirectory as denoted in the URL definition above (this is a recent config change and needs to be back-ported to 3.6). Currently (3.6) the default language code is published to a site that doesn't include language in the url (i.e. English is docs.darktable.org/usermanual/3.6/), so it might be nice to put some nginx config in place to automatically redirect (`try_files)` to `<lang>=en`.

If a page does not exist in a given language, attempt to redirect to the English version of the page (replace that part of the URL with `<lang>=en`).

Something like the following location setting will probably suffice to manage these redirects, though we might need another one to handle where no language code is provided (for example where the entire url provided is docs.darktable.org/development). My regex is rusty so I've probably missed something key (e.g. it currently will only match if at least two slashes appear following the version number)..

```
 location ~ ^(/usermanual/[^/]+)/([^/]+)/(.*)$ 
     { 
        try_files $uri $1/en/$2/$3 $1/en/$3 $1/en/$2/$3/index.html $1/en/$3/index.html; 
      }
```

**Comment by TurboGit** [comment]
Hope you don't mind, I have pinned this issue to ensure good visibility. I'm ready for the dt part as soon as all the links are in place.

**Comment by darix** [comment]
I dont understand to what would be different from what we have right now.

**Comment by elstoc** [comment]
I was trying to write a comprehensive (ish) set of requirements. Some of this is a restatement of what we have right now but some is new and some currently doesn't work properly, so this is also meant to define what we need to test.

The redirect to from the greatest version number to `/stable` is different, as is the use of `/development` instead of `/master`. I don't know how the auto-build works but there may also be some changes needed to that in order to pick up subsequent 3.6.x tags.

Some of this is changes for the dtdocs repo that we've already done (in master). The back-porting of the language stuff to 3.6 is still TBC.

**Comment by darix** [comment]
master is now published as development.

stable redirect is done. that is all?

**Comment by elstoc** [comment]
A few things from initial tests:

- I'm not sure whether we want the "untranslated" pages to just display the English version with the entered URL or to do a _redirect_ to EN (currently does the former). Probably we want @paperdigits and @TurboGit to weigh in here.
- I don't see the stable redirect. For example https://docs.darktable.org/usermanual/3.6/lighttable/ should redirect to https://docs.darktable.org/usermanual/stable/lighttable/ but doesn't. When we tag a 3.8.x release we will want _that_ to redirect instead. It would be great if we could do that without changing the nginx config again!
- In the 3.6 version some pages fail to load. It looks like it's just the first-level pages for each section. For example https://docs.darktable.org/usermanual/3.6/slideshow/ loads the home page but https://docs.darktable.org/usermanual/3.6/darkroom/pixelpipe/ works fine. This might be because the language settings on the built site don't match those on the development site. @paperdigits: I don't know how much of this stuff we want to port back to 3.6 at this point.
- If you don't include the trailing slash on the URL of an "untranslated" page you get a 404 error, for example https://docs.darktable.org/usermanual/development/xx/lighttable
- Without a page reference, "untranslated" pages do not redirect. For example https://docs.darktable.org/usermanual/development/xx/

**Comment by darix** [comment]
1. we did the redirect the other way around. stable points to 3.6.
2. the whole stuff would be a lot easier if we wouldnt need all those redirects to treat english special in 3.6
3. yes untranslated pages without a slash wont work

```
  location ~ ^(/usermanual/3.[^/]+)/([^/]+)/(.*)/$ {
    try_files $uri/index.html $1/$3/index.html;
  }
```

I mean one could add another rule that just says

```
  location ~ ^(/usermanual/3.[^/]+)/([^/]+)/([^/]+)$ {
    try_files $uri/index.html $1/$3/index.html;
  }
```

but that would only fix the first level. not sure we really really want to go with 

```
  location ~ ^(/usermanual/3.[^/]+)/([^/]+)/(.*)$ {
    try_files $uri/index.html $1/$3/index.html;
  }
```

at least backport the whole "render en to the subdirectory" having to have all those special redirect rules makes debugging this super annoying.

4. 404 is a very valid return code. not sure we really should add redirects for every stupid typo

**Comment by elstoc** [comment]
> we did the redirect the other way around. stable points to 3.6

Ideally I'd prefer most people to only ever see the `/stable` site so if they share links etc it doesn't point to an old version. This redirect is mainly for darktable help links, which always need to include the version number in the URL.

> the whole stuff would be a lot easier if we wouldnt need all those redirects to treat english special in 3.6

> at least backport the whole "render en to the subdirectory

Agreed. I'd like to create a 3.6.x branch in dtdocs and back-port some of those changes into it 

> 404 is a very valid return code. not sure we really should add redirects for every stupid typo

I get where you're coming from but since https://docs.darktable.org/usermanual/development/en/lighttable works, I would prefer if https://docs.darktable.org/usermanual/development/xx/lighttable did the same thing

**Comment by elstoc** [comment]
@TurboGit I think you can proceed with the darktable helplinks now using the summary I provided above for guidance. I think the links should all work for master - obviously we haven't published 3.8 yet but you should still be able to test that the URLs are correct (are consistent with the current development version).

For darktable master the help links should point to  `https://docs.darktable.org/usermanual/development/<lang>/path/to/page/` and for a release version it should point to `https://docs.darktable.org/usermanual/<ver>/<lang>/path/to/page/` where `<ver>` just uses the first two parts of the version number (e.g. 3.6, 3.8). In both cases `<lang>` should be the language code that darktable uses. You can see [here](https://github.com/darktable-org/dtdocs/tree/master/po) for the list of languages dtdocs supports (plus `/en` for the English version) but I suggest you just pass the language and don't worry about whether it's supported, since any unsupported language will just load the English page anyway.

**Comment by TurboGit** [comment]
Great! I'll look at this very soon. Thanks.

**Comment by elstoc** [comment]
I guess this can be closed now, since we have been using a consistent process for the last few versions.
